merkle.go

```diff
@@ -0,0 +1,189 @@
+		// the array can be shifted and shrunk
+		shiftAndShrinkArray(&nodes)
+	}
+
+	return &MerkleTree{
+		Root: nodes[0],
+	}, nil
+}
+
+// isValidDataSet makes sure the input set has elements
+func isValidDataSet(inputData [][]byte) bool {
+	return inputData != nil && len(inputData) > 0
+}
+
+// shiftAndShrinkArray shifts every other node to the
+// beginning of the array, and discards half of it (shrinks it).
+// Due to the way results are being stored (index of left child),
+// and the fact that the Merkle tree is a perfect binary tree,
+// it can be guaranteed that the results are on every other index in the node level array
+func shiftAndShrinkArray(nodes *[]*Node) {
+	// Put the results in the first half of the array.
+	// One counter keeps track of the next slot to place the value (moves by 1) (saveIndx)
+	// and the other keeps track of which element should be stored (moves by 2) (resultIndx)
+	initialLevelSize := len(*nodes)
+	saveIndx := 0
+	for resultIndx := 0; resultIndx < initialLevelSize; resultIndx += 2 {
+		(*nodes)[saveIndx] = (*nodes)[resultIndx]
+		saveIndx++
+	}
+
+	// Wipe the other half of the array, since
+	// all useful and needed results are in the first half
+	*nodes = (*nodes)[:initialLevelSize/2]
+}
+
+// generateLeaves generates the initial (leaf) level of the Merkle tree.
+// The leaf level needs to be a power of 2, since the Merkle tree is considered
+// to be a perfect binary tree
+func generateLeaves(inputData [][]byte, wp *workerPool) ([]*Node, error) {
+	inputDataSize := len(inputData)
+	leafLevelSize := inputDataSize
+
+	if inputDataSize > 1 {
+		// Find the nearest power of 2 for the base leaf level size
+		leafLevelSize = nearestPowerOf2(len(inputData))
+	}
+
+	leaves := make([]*Node, leafLevelSize)
+
+	// Create the initial job set for the leaf nodes,
+	// where each job is a single leaf node to be processed
+	for i := 0; i < inputDataSize; i++ {
+		wp.addJob(&workerJob{
+			storeIndex: i,
+			sourceData: [][]byte{
+				inputData[i],
+			},
+		})
+	}
+
+	// Grab the results from the worker pool
+	for i := 0; i < inputDataSize; i++ {
+		result := wp.getResult()
+		if result.error != nil {
+			return nil, fmt.Errorf(
+				"unable to perform hashing, %w",
+				result.error,
+			)
+		}
+
+		// Save the leaf nodes
+		leaves[i] = &Node{
+			Hash:   result.hashData,
+			Left:   nil,
+			Right:  nil,
+			Parent: nil,
+		}
+	}
+
+	// Since there is a possibility an expansion of the leaves array
+	// took place to make it a power of 2, the last element in the original leaf set
+	// needs to be duplicated to fill out the remaining (expanded) slots
+	lastNode := leaves[inputDataSize-1]
+	for i := inputDataSize; i < len(leaves); i++ {
+		leaves[i] = lastNode.duplicate()
+	}
+
+	return leaves, nil
+}
+
+// nearestPowerOf2 returns the nearest power of 2 to
+// the provided number.
+// Courtesy of BitTwiddlingHacks:
+// https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+func nearestPowerOf2(num int) int {
```

Comment:
nearest sounds an ambiguous little bit. the return value is always bigger or equals to given value, right? how about `nearestNextPowerOf2`? (I don't know it's more proper)

Reply:
You're right, it will return the bigger / equal to the given value.

I've added your suggestion:

[7cb1739](https://github.com/Trapesys/fastmerkle/pull/1/commits/7cb1739221c304730ae9cdd4a8b16732efa01961)